src/quicklist.c

```diff
@@ -67,20 +86,36 @@ int quicklistisSetPackedThreshold(size_t sz) {
  * This is used only if we're limited by record count. when we're limited by
  * size, the maximum limit is bigger, but still safe.
  * 8k is a recommended / default size limit */
+
+/* 任何存储多个元素的 listpack 的最大容量（字节）。
+ * 超出限制值的大元素将独立存储在 listpack 中。
+ * 这只在被元素数量限制的情况下使用（fill 为正数的情况）。
+ * 当我们被字节数限制时（fill 为负数的情况），最大的限制可以更大，且仍然安全。
+ * 8k 是一个推荐/默认的容量限制 */
 #define SIZE_SAFETY_LIMIT 8192
 
 /* Maximum estimate of the listpack entry overhead.
  * Although in the worst case(sz < 64), we will waste 6 bytes in one
  * quicklistNode, but can avoid memory waste due to internal fragmentation
  * when the listpack exceeds the size limit by a few bytes (e.g. being 16388). */
+
+/* listpack 条目（entry）开销的最大估计值。
+ * 尽管在最坏的情况下（sz < 64），我们将在一个快速列表节点中浪费6个字节，
+ * 但当 listpack 超过大小限制的几个字节时（例如：16388）可以避免由于内部碎片造成的内存浪费 */
```

Comment:
@enjoy-binbin 我想在这里加个例子说明的，但是我实在想不出来为什么会出现这些情况...你知道吗？（最近学校课设太多没什么时间更新...先更新个100行吧..）
```
" Although in the worst case(sz < 64), we will waste 6 bytes in one
  quicklistNode, but can avoid memory waste due to internal fragmentation
  when the listpack exceeds the size limit by a few bytes (e.g. being 16388). "
```

Reply:
我不太熟悉 listpack... 我来看看叫个外援 @sundb 哈哈哈有空阔以帮我们描述一下么

Reply:
最近没在电脑前面，等我晚上有电脑了我再看看。

Reply:
哦哦，我还在想为什么要做额外估算，原来是往最大的header和backlen长度估算了，多谢指教🙏

Reply:
> 因为qucklistNode的最大长度可以配置成-5(即16k)
```
static const size_t optimization_level[] = {4096, 8192, 16384, 32768, 65536};
```
这样的说法是否有点奇怪？或者是您其实省略了解释吗？配置成负数时使用的是节点容量的限制，-5的时候最大限制是64k，但是因为count字段只有16位，即使节点存储的元素每个都是最小的1B，所以它最多也不能超过16k。

Reply:
16k是我说错了，-5的时候是64k，它的最大header和backlen也是8.
listpack的长度存储只有16位不代表它只能存储16k的数据，到超过这个限制时listpack的长度过去每次都需要遍历，非常慢。

Reply:
嗯……是这样，但是我指的是quicklistnode里的count字段，这个count是16位的。listpack可以超过16k然后遍历全部来获取元素数量，我想知道quicklistnode是不是也这样。但是我去看了看PushHead和PushTail的函数，head->count和tail->count好像是直接++，并没有会不会溢出的判断？虽然不推荐设置32k和64k的限制，但是如果真的设置了并且每个节点元素超过16k个，这样的话会不会出现溢出问题？
（不在电脑前不方便复制源码到评论上，抱歉🙏）

Reply:
我印象中 quicklist 不是有个 allowinsert 啥的吗，如果超过限制就会创建一个新的节点作为新的头或者新的尾，所以 count++ 是 ok 的

Reply:
是有这个，但是看起来并不会处理这个问题？（用手机试着复制一下代码，不知道会不会正常…）
看起来只是判断了满不满足fill所描述的限制内，满足就可以插入。
```
 
 ​REDIS_STATIC ​int​ ​_quicklistNodeAllowInsert​(​const​ quicklistNode *node, 
 ​                                           ​const​ ​int​ fill, ​const​ ​size_t​ sz) { 
 ​    ​if​ (​unlikely​(!node)) 
 ​        ​return​ ​0​; 
  
 ​    ​if​ (​unlikely​(​QL_NODE_IS_PLAIN​(node) || ​isLargeElement​(sz))) 
 ​        ​return​ ​0​; 
  
 ​    ​/*​ Estimate how many bytes will be added to the listpack by this one entry. 
 ​     * We prefer an overestimation, which would at worse lead to a few bytes 
 ​     * below the lowest limit of 4k (see optimization_level). 
 ​     * Note: No need to check for overflow below since both `node->sz` and 
 ​     * `sz` are to be less than 1GB after the plain/large element check above. ​*/ 
 ​    ​size_t​ new_sz = node->​sz​ + sz + SIZE_ESTIMATE_OVERHEAD; 
 ​    ​if​ (​likely​(​_quicklistNodeSizeMeetsOptimizationRequirement​(new_sz, fill))) 
 ​        ​return​ ​1​; 
 ​    ​/*​ when we return 1 above we know that the limit is a size limit (which is 
 ​     * safe, see comments next to optimization_level and SIZE_SAFETY_LIMIT) ​*/ 
 ​    ​else​ ​if​ (!​sizeMeetsSafetyLimit​(new_sz)) 
 ​        ​return​ ​0​; 
 ​    ​else​ ​if​ ((​int​)node->​count​ < fill) 
 ​        ​return​ ​1​; 
 ​    ​else 
 ​        ​return​ ​0​; 
 ​}
```

Reply:
fill有最大值限制，quicklistnode的count不会溢出

Reply:
```c
 ​    ​size_t​ new_sz = node->​sz​ + sz + SIZE_ESTIMATE_OVERHEAD; 
     // new_sz 元素插入后的节点新大小，是否满足 fill 配置项，满足的话则允许在节点中插入，否则不允许
    //  这里面应该是限制了 new_sz 的大小是不会超过对应 fill 的配置项，例如 64K 啥的
    //  _quicklistNodeSizeMeetsOptimizationRequirement 里 fill >= 0 会直接 return
 ​    ​if​ (​likely​(​_quicklistNodeSizeMeetsOptimizationRequirement​(new_sz, fill))) 
 ​        ​return​ ​1​; 
 ​    ​/*​ when we return 1 above we know that the limit is a size limit (which is 
 ​     * safe, see comments next to optimization_level and SIZE_SAFETY_LIMIT) ​*/ 
    // new_sz 不满足安全界限，规定 new_sz 要 <= 8k，限制单个 listpack 最大到 8KB
    // 当然这里只是说不给往这个 quicklistNode 中插入新 entry，但是是可以单独在一个节点中保存 >8KB 的 listpack的
 ​    ​else​ ​if​ (!​sizeMeetsSafetyLimit​(new_sz)) 
 ​        ​return​ ​0​; 
     // 到这里的话，fill 是正数的情况，判断 node->count + 1 <= fill，控制单个 listpack 里 entry 的数量
 ​    ​else​ ​if​ ((​int​)node->​count​ < fill) 
 ​        ​return​ ​1​; 
 ​    ​else 
 ​        ​return​ ​0​; 
```

fill 是有限制的
```
#define FILL_MAX ((1 << (QL_FILL_BITS-1))-1)
void quicklistSetFill(quicklist *quicklist, int fill) {
    if (fill > FILL_MAX) {
        fill = FILL_MAX;
    } else if (fill < -5) {
        fill = -5;
    }
    quicklist->fill = fill;
}
```

Reply:
fill为正数的时候有最大值限制，fill为负数的时候有size的限制，之前忘记了这个size限制，这下就了解了，多谢🙏🙏

Reply:
```
/* Optimization levels for size-based filling.
 * Note that the largest possible limit is 16k, so even if each record takes
 * just one byte, it still won't overflow the 16 bit count field. */
```
感觉上午自己一直在被这个16k所误导，这里是注释错误吗？2^16应该是64k吧？😢

Reply:
好的感谢，那我顺便帮原注释更新一下